tidb-binlog/pump_client/client.go

```diff
@@ -255,89 +262,157 @@ func (c *PumpsClient) WriteBinlog(binlog *pb.Binlog) error {
 			err = errors.New(resp.Errmsg)
 		}
 		if err == nil {
+			successPump = pump
 			return nil
 		}
-		Logger.Errorf("[pumps client] write binlog error %v", err)
+		errNums[pump]++
+		Logger.Warnf("[pumps client] write binlog to pump %s (type: %s, start ts: %d, commit ts: %d, length: %d) error %v", pump.NodeID, binlog.Tp, binlog.StartTs, binlog.CommitTs, len(commitData), err)
 
 		if binlog.Tp != pb.BinlogType_Prewrite {
 			// only use one pump to write commit/rollback binlog, util write success or blocked for ten minutes. And will not return error to tidb.
-			if time.Since(startTime) > CommitBinlogMaxRetryTime {
-				Logger.Warnf("[pumps client] write commit binlog %d failed, error %v", binlog.CommitTs, err)
-				return nil
+			if time.Since(startTime) > CommitBinlogTimeout {
+				break
 			}
 		} else {
 			if !isRetryableError(err) {
 				// this kind of error is not retryable, return directly.
 				return err
 			}
 
-			// every pump can retry 5 times, if retry 5 times and still failed, set this pump unavaliable, and choose a new pump.
-			if (retryTime+1)%RetryTime == 0 {
-				c.setPumpAvaliable(pump, false)
-				pump = c.Selector.Next(binlog, retryTime/5+1)
-				Logger.Debugf("[pumps client] avaliable pumps: %v, write binlog choose pump %v", c.Pumps.AvaliablePumps, pump)
+			if time.Since(startTime) > c.BinlogWriteTimeout {
+				break
 			}
 
 			retryTime++
-			if retryTime > c.RetryTime {
-				break
+		}
+
+		if binlog.Tp != pb.BinlogType_Prewrite {
+			time.Sleep(RetryInterval * 10)
+		} else {
+			time.Sleep(RetryInterval)
+		}
+	}
+
+	pump, err1 := c.backoffWriteBinlog(req, binlog.Tp)
+	if err1 == nil {
+		successPump = pump
+		return nil
+	}
+
+	return errors.Errorf("write binlog failed, the last error %v", err)
+}
+
+func (c *PumpsClient) backoffWriteBinlog(req *pb.WriteBinlogReq, binlogType pb.BinlogType) (pump *PumpStatus, err error) {
+	if binlogType != pb.BinlogType_Prewrite {
+		// never return error for commit/rollback binlog.
+		return nil, nil
+	}
+
+	unAvaliablePumps := make([]*PumpStatus, 0, 3)
+	c.Pumps.RLock()
+	for _, pump := range c.Pumps.UnAvaliablePumps {
+		unAvaliablePumps = append(unAvaliablePumps, pump)
+	}
+	c.Pumps.RUnlock()
+
+	var resp *pb.WriteBinlogResp
+	// send binlog to unavaliable pumps to retry again.
+	for _, pump := range unAvaliablePumps {
+		resp, err = pump.WriteBinlog(req, c.BinlogWriteTimeout)
+		if err == nil {
+			if resp.Errmsg != "" {
+				err = errors.New(resp.Errmsg)
+			} else {
+				// if this pump can write binlog success, set this pump to avaliable.
+				Logger.Debugf("[pumps client] write binlog to unavaliable pump %s success, set this pump to avaliable", pump.NodeID)
+				c.setPumpAvaliable(pump, true)
+				return pump, nil
+			}
+		}
+	}
+
+	return nil, err
+}
+
+func (c *PumpsClient) addErrNum(nodeID string, num int64) {
+	c.Pumps.Lock()
+	c.Pumps.ErrNums[nodeID] += num
+	c.Pumps.Unlock()
+}
+
+func (c *PumpsClient) clearErrNum(nodeID string) {
+	c.Pumps.RLock()
```

Comment:
don't use two lock, remove L361 ~ L366

Reply:
in most time will write binlog success, so the node id is not exists in ErrNums map, so I use a read lock here first.

Reply:
put it into pump self attributes，and use atomic operation